### ISSUE TITLE
Setup fails on HP Tape drive : Inappropriate ioctl for device

### ISSUE THREAD

**Issue by mpideil** [ansible/ansible#14068]
Hello,

I'l using Ansible 2.0.0.2

The setup module fails on HP SAS controller. 

```
# ansible SERVER35678 -m setup -vvv
Using /etc/ansible/ansible.cfg as config file
<---snip--->
SERVER35678 | FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "setup"
    },
    "module_stderr": "Killed by signal 1.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/.ansible/xxx/setup\", line 5142, in ?\r\n    main()\r\n  File \"/tmp/.ansiblexxx/setup\", line 140, in main\r\n    data = run_setup(module)\r\n  File \"/tmp/.ansible/xxx/setup\", line 84, in run_setup\r\n    facts = ansible_facts(module)\r\n  File \"/tmp/.ansible/xxx/setup\", line 5078, in ansible_facts\r\n    facts.update(Hardware().populate())\r\n  File \"/tmp/.ansible/xxx/setup\", line 2899, in populate\r\n    self.get_device_facts()\r\n  File \"/tmp/.ansible/xxx/setup\", line 3179, in get_device_facts\r\n    d[key] = get_file_content(sysdir + \"/device/\" + key)\r\n  File \"/tmp/.ansible/xxx/setup\", line 5057, in get_file_content\r\n    data = datafile.read()\r\nIOError: [Errno 25] Inappropriate ioctl for device\r\n",
    "msg": "MODULE FAILURE",
    "parsed": false
}
```

```
# python -m trace --trace /tmp/.ansible/xxx/setup
< --- snip --- >
subprocess.py(984):             if errwrite and errread:
subprocess.py(985):                 os.close(errwrite)
subprocess.py(988):             data = os.read(errpipe_read, 1048576) # Exceptions limited to 1 MB
<blocked>
```

```
# cat /sys/block/cciss\!c1d0/device/vendor
cat: /sys/block/cciss!c1d0/device/vendor: Inappropriate ioctl for device
```

The controller is a bot tape drive connected to a HP Smart Array Controller.
It is recognize by the system though :

```
# cat /proc/driver/cciss/cciss1
cciss1: HP Smart Array P212 Controller
Board ID: 0x3241103c
Firmware Version: 2.72
IRQ: 170
Logical drives: 0
Current Q depth: 0
Current # commands on controller: 0
Max Q depth since init: 0
Max # commands on controller since init: 3
Max SG entries since init: 0
Sequential access devices: 1
```

​

```
# hpacucli ctrl all show config detail​
Smart Array P212 in Slot 4
   Bus Interface: PCI
   Slot: 4
   Serial Number: PACCP9SYCKPB
   RAID 6 (ADG) Status: Disabled
   Controller Status: OK
   Chassis Slot:
   Hardware Revision: Rev C
   Firmware Version: 2.72
   Rebuild Priority: Medium
   Expand Priority: Medium
   Surface Scan Delay: 0 secs
   Wait for Cache Room: Disabled
   Surface Analysis Inconsistency Notification: Disabled
   Post Prompt Timeout: 0 secs
   Cache Board Present: False
   Drive Write Cache: Disabled
   SATA NCQ Supported: True
```

Any hints ?

Cheers
Matt

**Comment by chrrrles** [comment]
Hello - What operating systems are involved, and also do you know if the `setup` from 1.9 works with this machine?

**Comment by mpideil** [comment]
Hello, thanks for your help.

**Issue Type:**
Bug Report

**Ansible Version:**
(Same issue with 1.9)
$ ansible --version
ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

**Ansible Configuration:**

```
[defaults]
remote_tmp     = /tmp/.ansible/tmp
forks          = 10
[ssh_connection]
ssh_args =
```

**Environment:**
Controller node : RHEL 7
Managed node : RHEL 5.4 (No issues on RHEL 5.6)

**Summary:**
The **setup** _module_ fail on read '/sys/block/cciss!c1d0/device/vendor'
cat: /sys/block/cciss!c1d0/device/vendor: Inappropriate ioctl for device​

The device is a HP Smart Array P212 with a tape robot
Others HPSmartArray answer correctly

**Steps To Reproduce:**
$ ansible -m setup mybadnode

**Expected Results:**
Successful execution 

**Actual Results:**

```
mybadnode | FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "setup"
    },
    "module_stderr": "Killed by signal 1.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 5142, in ?\r\n    main()\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 140, in main\r\n    data = run_setup(module)\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 84, in run_setup\r\n    facts = ansible_facts(module)\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 5078, in ansible_facts\r\n    facts.update(Hardware().populate())\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 2899, in populate\r\n    self.get_device_facts()\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 3179, in get_device_facts\r\n    d[key] = get_file_content(sysdir + \"/device/\" + key)\r\n  File \"/tmp/.ansible/tmp/ansible-tmp-xxx/setup\", line 5057, in get_file_content\r\n    data = datafile.read()\r\nIOError: [Errno 25] Inappropriate ioctl for device\r\n",
    "msg": "MODULE FAILURE",
    "parsed": false
}
```

**Details:**
_RHEL 5.4 HP Smart Array P212_
$ find /sys/block/cciss!c1d0/device/ -exec cat {} \;
cat: /sys/block/cciss!c1d0/device/: Is a directory
cat: /sys/block/cciss!c1d0/device/block:cciss!c1d0: Is a directory
cat: /sys/block/cciss!c1d0/device/lunid: Inappropriate ioctl for device
ccisscat: /sys/block/cciss!c1d0/device/unique_id: Inappropriate ioctl for device
cat: /sys/block/cciss!c1d0/device/rev: Inappropriate ioctl for device
cat: /sys/block/cciss!c1d0/device/model: Inappropriate ioctl for device
cat: /sys/block/cciss!c1d0/device/vendor: Inappropriate ioctl for device
0
   0.00GB
RAID UNKNOWN
cat: /sys/block/cciss!c1d0/device/power: Is a directory

0
cat: /sys/block/cciss!c1d0/device/uevent: Permission denied
​​

_RHEL 5.6 HP Smart Array P212_
$ find /sys/block/cciss!c1d0/device/ -exec cat {} \;
cat: /sys/block/cciss!c1d0/device/: Is a directory
cat: /sys/block/cciss!c1d0/device/block:cciss!c1d0: Is a directory
0
RAID UNKNOWN
cat: /sys/block/cciss!c1d0/device/lunid: Inappropriate ioctl for device

00000000000000000000000000000000
cat: /sys/block/cciss!c1d0/device/bus: Is a directory
cat: /sys/block/cciss!c1d0/device/subsystem: Is a directory
cat: /sys/block/cciss!c1d0/device/power: Is a directory

0
cat: /sys/block/cciss!c1d0/device/uevent: Permission denied

**Patch: (I hope it will help)**
/usr/lib/python2.7/site-packages/ansible/module_utils/facts.py​:2973

```
def get_file_content(path, default=None, strip=True):
    data = default
    if os.path.exists(path) and os.access(path, os.R_OK):
        try:
            datafile = open(path)
            data = datafile.read()   <---- IOError not caught
            if strip:
                data = data.strip()
            if len(data) == 0:
                data = default
        finally:
            datafile.close()
    return data
```

**Comment by bcoca** [comment]
fixed via 2ec31c2c7981ad18d3e42a3a37e23ab1810cd31b which is now in 2.0.1
